### DISCUSSION TITLE
Opcode 30、57、58 for DS

### DISCUSSION THREAD

**Discussion by Yzgll** [sarchlab/mgpusim#62] [Q&A]
Teacher Sun, hello, I was recently researching the emulator's local memory, and I wanted to load some global memory data into local memory for use, so I added three new instructions. `DS_WRITE_B8: DS[A] = D0[7:0];  byte write`, `DS_READ_I8: R = signext(DS[A][7:0]}; signed byte read `and `DS_READ_U8: R = {24 'h0,DS[A][7:0]}; unsigned byte read`. However, after this I got the wrong result, which was found to be due to the inconsistency between the data loaded into local memory and the data in global memory. Therefore, I suspect that the three instruction functions added are not correct, but I have checked for several times but could not find the error, so I would like to ask you to help me check where the error is. Looking forward to your reply.

```
func (u *ALUImpl) runDSWRITEB8(state InstEmuState) {
	inst := state.Inst()
	sp := state.Scratchpad()
	layout := sp.AsDS()
	lds := u.LDS()

	i := uint(0)
	for i = 0; i < 64; i++ {
		if !laneMasked(layout.EXEC, i) {
			continue
		}

		addr := layout.ADDR[i] + inst.Offset0
		data0offset := uint(8 + 64*4)
		byteToWrite := sp[data0offset+i*16]
		lds[addr] = byteToWrite
	}
}
```
```
func (u *ALUImpl) runDSREADI8(state InstEmuState) {
	inst := state.Inst()
	sp := state.Scratchpad()
	layout := sp.AsDS()
	lds := u.LDS()

	i := uint(0)
	for i = 0; i < 64; i++ {
		if !laneMasked(layout.EXEC, i) {
			continue
		}

		addr0 := layout.ADDR[i] + inst.Offset0
		bytedata := lds[addr0]
		signedData := int32(int8(bytedata)&0xFF)
		dstOffset := uint(8 + 64*4 + 256*4*2)
		sp[dstOffset] = byte(signedData)
		sp[dstOffset+1] = byte(signedData >> 8)
		sp[dstOffset+2] = byte(signedData >> 16)
		sp[dstOffset+3] = byte(signedData >> 24)

	}
}
```
```
func (u *ALUImpl) runDSREADU8(state InstEmuState) {
	inst := state.Inst()
	sp := state.Scratchpad()
	layout := sp.AsDS()
	lds := u.LDS()
	i := uint(0)
	for i = 0; i < 64; i++ {
		if !laneMasked(layout.EXEC, i) {
			continue
		}
		addr0 := layout.ADDR[i] + inst.Offset0
		byteData := lds[addr0]
		usignedData := uint32(uint8(byteData)0xFF)
		dstOffset := uint(8 + 64*4 + 256*4*2)
		sp[dstOffset] = byte(usignedData)
		sp[dstOffset+1] = byte(usignedData >> 8)
		sp[dstOffset+2] = byte(usignedData >> 16)
		sp[dstOffset+3] = byte(usignedData >> 24)

	}
}
```

**Comment by syifan** [comment]
I do not understand how you added new instructions. How can you make the AMD compilers generate the instructions? Can I see the assembly?

**Reply by Yzgll** [reply]
I just added case 30, case 57, and case 58 to aluds.go, and implemented the corresponding functions runDSWRITEB8(state), runDSREADI8(state), and runDSREADU8(state) below to simulate the instruction. And when you mentioned "assembly", were you referring to kernels.hsaco?

**Reply by syifan** [reply]
Or what is your OpenCL program? How do you know the added instructions are used in the hsaco file?

**Reply by Yzgll** [reply]
Here is the kernel of my OpenCL program, with comments added where I made modifications. After modifying the kernel and before adding the instructions, the program panicked and could not run, displaying the message: "Opcode 30, 57 for DS format is not implemented". After I added the instructions, the program was able to run successfully, but the computation results were incorrect. Therefore, it seems to me that these instructions are used in the hsaco file, and the computation errors are due to mistakes in the code I wrote.
![1](https://github.com/sarchlab/mgpusim/assets/165646862/272d3b36-95e0-4462-9428-01e1485649aa)

```
__kernel void Encrypt(__global uchar* input, 
                      __global uint* expanded_key, 
                      __global uchar* s) {
  uchar state[16];
  int tid = get_global_id(0);
  
  //load s_box to local memory
  __local uchar local_s[256];
  for(int i = 0;i < 256;i ++){
    local_s[i]=s[i];
  }

  for (int i = 0; i < 16; i++) {
    state[i] = input[tid * 16 + i];
  }

  AddRoundKey(state, expanded_key, 0);
  for (int i = 1; i < 10; i++) {
    SubBytes(state, local_s);//use local_s not s
    ShiftRows(state);
    MixColumns(state);
    AddRoundKey(state, expanded_key, i * 4);
  }
  SubBytes(state, local_s);
  ShiftRows(state);
  AddRoundKey(state, expanded_key, 10 * 4);

  for (int i = 0; i < 16; i++) {
    input[tid * 16 + i] = state[i];
  }
}

```

**Comment by Yzgll** [comment]
> Or what is your OpenCL program? How do you know the added instructions are used in the hsaco file?

You are right.  I set the input as fixed and unchangeable.  No matter how I modify the three added instruction functions, the resulting output is still incorrect and does not change with the modifications to the instruction functions.  I even set the added instruction functions as empty functions and still got the same error result.  It seems that the instruction functions I added only allow the program to run but do not implement the corresponding functionality.  Why is this happening?

**Reply by syifan** [reply]
The quick way to verify if the instructions are used is to add prints to the emulation function and see if the prints generate output. You can use the compiler with `-S` argument to generate the assembly file to check if the instructions are used.

**Reply by Yzgll** [reply]
Thanks. I'll try it tomorrow. One more issue, I had a problem with ISA Debugger and the 64 debug files I generated were all empty!
![3](https://github.com/sarchlab/mgpusim/assets/165646862/5c364345-712b-43c9-8de5-5206bb573386)
![4](https://github.com/sarchlab/mgpusim/assets/165646862/364a3122-11ad-4c0a-be7d-54f399759fa1)
![5](https://github.com/sarchlab/mgpusim/assets/165646862/9dd7c2ec-b7cd-4164-af63-bc070408faab)

**Reply by Yzgll** [reply]
> The quick way to verify if the instructions are used is to add prints to the emulation function and see if the prints generate output. You can use the compiler with `-S` argument to generate the assembly file to check if the instructions are used.

I found these in the assembly file. This is enough to prove that the instructions runDSWRITEB8 and runDSREADI8 that I added were actually used, and that the calculation was wrong because I wrote the wrong code. So I should next modify the runDSWRITEB8 and runDSREADI8 functions through unit testing, right?
![33](https://github.com/sarchlab/mgpusim/assets/165646862/1fa6322e-6f4e-4c15-aab6-ed323530090a)
![22](https://github.com/sarchlab/mgpusim/assets/165646862/561921ea-35dc-43c2-a52f-2140385e1cbb)

**Reply by syifan** [reply]
Yes. If you think your emulation logic is wrong, you should change both your unit test and the production code. 

BTW, before you work on changing the implementation, you should consider using MGPUSim to disassemble the hsaco to check if MGPUSim can generate the same instruction string as the official compiler. The disassembler can be found here https://github.com/sarchlab/mgpusim/tree/v3/insts/gcn3disassembler.

**Reply by syifan** [reply]
For the isa debugger, the problem is here https://github.com/sarchlab/mgpusim/blob/v3/emu/isadebugger.go#L40. Uncomment these lines should help. You may want to use a similar condition to prevent the isadebugger from generating too long traces.

**Reply by Yzgll** [reply]
> these lines should help. You may want to use a similar condition to prevent the isadebugger from generating too long traces.

I tried this today, but I still had some trouble. First of all, after uncommenting, there are two functions that do not define h.logDiffWf(wf) and h.stubWf(wf), do I need to add these two functions myself? Second, I simply implemented these two functions and finally successfully generated a.debug file. But there was an error after renaming to a json file, and the browser still saw nothing. Third, I don't understand the relationship between a test file like aluds_test.go and isaDebugger, can I test only for the test file I want?
![QQ图片20240528212538](https://github.com/sarchlab/mgpusim/assets/165646862/62f3bfab-48d8-4a03-ad0e-398cd6379f33)

**Reply by syifan** [reply]
Do not use the stub function. Use the `logWholeWF` instead. 

Test files are for tests, and they are not part of the MGPUSim simulations. If you want to run tests, you can run `go test` in the folder that contains the test.

**Reply by Yzgll** [reply]
As you said, I uncommented and only used `logWholeWF`, but still nothing on the browser side. I just followed the steps in `readme` to rename the `debug `file and start the http server. The second problem is that `go test` can not run the test file, showing something undefined, why?
![3](https://github.com/sarchlab/mgpusim/assets/165646862/6f4b7984-4f99-4609-a3f9-9218fa58f05f)
![1](https://github.com/sarchlab/mgpusim/assets/165646862/012e7860-d954-44da-8fa7-5af53750efb8)
![2](https://github.com/sarchlab/mgpusim/assets/165646862/46fc54ae-9636-400d-8965-1bf9713c93bc)

**Reply by syifan** [reply]
Can you try removing the two if statements and just keep the `h.logWholeWf`?

**Reply by Yzgll** [reply]
Can I interpret it as a direct call to h.logWholeWf(wf)?

**Reply by syifan** [reply]
Yes.
